mbrl/planning/trajectory_opt.py

```diff
@@ -153,6 +153,327 @@ def optimize(
+            )
+            population = torch.where(
+                population < self.lower_bound, self.lower_bound, population
+            )
+            values = obj_fun(population)
+            values[values.isnan()] = -1e-10
+
+            if callback is not None:
+                callback(population, values, k)
+
+            # weight actions
+            weights = torch.reshape(
+                torch.exp(self.gamma * (values - values.max())),
+                (self.population_size, 1, 1),
+            )
+            norm = torch.sum(weights) + 1e-10
+            weighted_actions = population * weights
+            self.mean = torch.sum(weighted_actions, dim=0) / norm
+
+        return self.mean.clone()
+
+
+class ICEMOptimizer(Optimizer):
+    """Implements the Improved Cross-Entropy Method (iCEM) optimization algorithm.
+
+    iCEM improves the sample efficiency over standard CEM and was introduced by
+    [2] for real-time planning.
+
+    Args:
+        num_iterations (int): the number of iterations (generations) to perform.
+        elite_ratio (float): the proportion of the population that will be kept as
+            elite (rounds up).
+        population_size (int): the size of the population.
+        population_decay_factor (float): fixed factor for exponential decrease in population size
+        colored_noise_exponent (float): colored-noise scaling exponent for generating correlated
+            action sequences.
+        lower_bound (sequence of floats): the lower bound for the optimization variables.
+        upper_bound (sequence of floats): the upper bound for the optimization variables.
+        keep_elite_frac (float): the fraction of elites to keep (or shift) during CEM iterations
+        alpha (float): momentum term.
+        device (torch.device): device where computations will be performed.
+        return_mean_elites (bool): if ``True`` returns the mean of the elites of the last
+            iteration. Otherwise, it returns the max solution found over all iterations.
+        round_population (bool): if "True" round population size to be a multiple of ensemble size
+            (must be True for Gaussian MLP ensemble)
+        ensemble_size (int): dynamics model ensemble size
+
+    [2] C. Pinneri, S. Sawant, S. Blaes, J. Achterhold, J. Stueckler, M. Rolinek and
+    G, Martius, Georg. "Sample-efficient Cross-Entropy Method for Real-time Planning".
+    Conference on Robot Learning, 2020.
+    """
+
+    def __init__(
+        self,
+        num_iterations: int,
+        elite_ratio: float,
+        population_size: int,
+        population_decay_factor: float,
+        colored_noise_exponent: float,
+        lower_bound: Sequence[Sequence[float]],
+        upper_bound: Sequence[Sequence[float]],
+        keep_elite_frac: float,
+        alpha: float,
+        device: torch.device,
+        return_mean_elites: bool = False,
+        round_population: bool = False,
+        ensemble_size: int = None,
```

Comment:
It would be better to remove `ensemble_size` and `round_population` and replace them with a single kwarg like `population_size_modulo: Optional[int] = None`, which does nothing if it's `None`, otherwise does the rounding. There is probably a better name for this than the one I'm suggesting, but hopefully the idea is clear.